File: tests/tool_dll_test.py
import glob
import os
import shutil
import subprocess
import sys
import tempfile
import unittest


def _compiler():
    return os.environ.get("CXX") or shutil.which("g++") or shutil.which("clang++")


def _shared_lib_ext():
    if sys.platform.startswith("win"):
        return ".dll"
    if sys.platform == "darwin":
        return ".dylib"
    return ".so"


class ToolDllTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.repo_root = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
        cls.compiler = _compiler()
        if not cls.compiler:
            raise unittest.SkipTest("No C++ compiler available")
        cls.serialized = sorted(
            glob.glob(os.path.join(cls.repo_root, "module_library", "serialized", "*.gpmod"))
        )
        if not cls.serialized:
            raise unittest.SkipTest("No serialized modules to test")

    def test_serializations_to_tool_dlls(self):
        with tempfile.TemporaryDirectory(prefix="tool_dll_test_") as temp_root:
            modlib_path = os.path.join(temp_root, "module_library.txt")
            source_root = os.path.join(temp_root, "module_library")
            os.makedirs(source_root, exist_ok=True)

            module_lines = ["MODULELIB V1", f"ROOT \"{source_root}\""]
            for idx, ser_path in enumerate(self.serialized):
                module_id = os.path.splitext(os.path.basename(ser_path))[0]
                tool_id = f"tool_{module_id}"
                module_lines.append(
                    "MODULE {id} {idx} \"{label}\" \"{serialized}\" \"\" 1 \"{tool_id}\" 0".format(
                        id=module_id,
                        idx=idx,
                        label=module_id,
                        serialized=os.path.abspath(ser_path),
                        tool_id=tool_id,
                    )
                )

            with open(modlib_path, "w", encoding="utf-8") as f:
                f.write("\n".join(module_lines))
                f.write("\n")

            generator = os.path.join(temp_root, "modlib_actualize")
            generator_cpp = os.path.join(temp_root, "modlib_actualize.cpp")
            with open(generator_cpp, "w", encoding="utf-8") as f:
                f.write(
                    "#include \"module_library_actualizer.h\"\n"
                    "#include <iostream>\n\n"
                    "int main(int argc, char** argv) {\n"
                    "    if (argc < 3) {\n"
                    "        std::cerr << \"usage: modlib_actualize <lib> <root>\\n\";\n"
                    "        return 2;\n"
                    "    }\n"
                    "    return gp_module_library_actualize_from_file(argv[1], argv[2]) ? 0 : 1;\n"
                    "}\n"
                )

            compile_cmd = [
                self.compiler,
                "-std=c++17",
                "-I",
                self.repo_root,
                generator_cpp,
                os.path.join(self.repo_root, "module_library_actualizer.cpp"),
                os.path.join(self.repo_root, "module_library.cpp"),
                "-o",
                generator,
            ]
            subprocess.run(compile_cmd, check=True)

            subprocess.run([generator, modlib_path, source_root], check=True)

            loader = os.path.join(temp_root, "tool_loader")
            loader_cpp = os.path.join(temp_root, "tool_loader.cpp")
            with open(loader_cpp, "w", encoding="utf-8") as f:
                f.write(
                    "#include \"tool_api.h\"\n"
                    "#include <iostream>\n"
                    "#include <memory>\n"
                    "#include <string>\n"
                    "#if defined(_WIN32)\n"
                    "#include <windows.h>\n"
                    "#else\n"
                    "#include <dlfcn.h>\n"
                    "#endif\n\n"
                    "class StubHost : public HostAPI {\n"
                    "public:\n"
                    "    void log(const char* message) override {\n"
                    "        if (message) { std::cerr << message << \"\\n\"; }\n"
                    "    }\n"
                    "    void draw_text(const TextRenderArgs& args) override { (void)args; }\n"
                    "};\n\n"
                    "int main(int argc, char** argv) {\n"
                    "    if (argc < 2) {\n"
                    "        std::cerr << \"usage: tool_loader <dll> [serialized]\\n\";\n"
                    "        return 2;\n"
                    "    }\n"
                    "    const char* path = argv[1];\n"
                    "    const char* ser = (argc > 2) ? argv[2] : nullptr;\n"
                    "#if defined(_WIN32)\n"
                    "    HMODULE handle = LoadLibraryA(path);\n"
                    "    if (!handle) {\n"
                    "        std::cerr << \"LoadLibrary failed\\n\";\n"
                    "        return 3;\n"
                    "    }\n"
                    "    auto create = reinterpret_cast<CreateToolFn>(GetProcAddress(handle, \"create_tool\"));\n"
                    "    if (!create) {\n"
                    "        std::cerr << \"create_tool not found\\n\";\n"
                    "        FreeLibrary(handle);\n"
                    "        return 4;\n"
                    "    }\n"
                    "#else\n"
                    "    void* handle = dlopen(path, RTLD_NOW);\n"
                    "    if (!handle) {\n"
                    "        std::cerr << dlerror() << \"\\n\";\n"
                    "        return 3;\n"
                    "    }\n"
                    "    auto create = reinterpret_cast<CreateToolFn>(dlsym(handle, \"create_tool\"));\n"
                    "    if (!create) {\n"
                    "        std::cerr << \"create_tool not found\\n\";\n"
                    "        dlclose(handle);\n"
                    "        return 4;\n"
                    "    }\n"
                    "#endif\n"
                    "    std::unique_ptr<ITool> tool(create());\n"
                    "    if (!tool) {\n"
                    "        std::cerr << \"create_tool returned null\\n\";\n"
                    "        return 5;\n"
                    "    }\n"
                    "    ToolInitContext ctx{};\n"
                    "    ctx.serialized_path = ser;\n"
                    "    StubHost host;\n"
                    "    RenderContext rctx{};\n"
                    "    tool->initialize(ctx);\n"
                    "    tool->tick(0.016, host);\n"
                    "    tool->render(rctx);\n"
                    "    tool->shutdown();\n"
                    "#if defined(_WIN32)\n"
                    "    FreeLibrary(handle);\n"
                    "#else\n"
                    "    dlclose(handle);\n"
                    "#endif\n"
                    "    return 0;\n"
                    "}\n"
                )

            compile_cmd = [
                self.compiler,
                "-std=c++17",
                "-I",
                self.repo_root,
                loader_cpp,
                "-o",
                loader,
            ]
            if not sys.platform.startswith("win"):
                compile_cmd.append("-ldl")
            subprocess.run(compile_cmd, check=True)

            tool_sources = glob.glob(os.path.join(source_root, "source", "tools", "*.cpp"))
            self.assertTrue(tool_sources, "No tool sources generated")

            for tool_source in tool_sources:
                tool_name = os.path.splitext(os.path.basename(tool_source))[0]
                tool_lib = os.path.join(temp_root, tool_name + _shared_lib_ext())
                build_cmd = [
                    self.compiler,
                    "-std=c++17",
                    "-shared",
                    "-fPIC",
                    "-I",
                    self.repo_root,
                    tool_source,
                    "-o",
                    tool_lib,
                ]
                if sys.platform.startswith("win"):
                    build_cmd = [
                        self.compiler,
                        "-std=c++17",
                        "-shared",
                        "-I",
                        self.repo_root,
                        tool_source,
                        "-o",
                        tool_lib,
                    ]
                subprocess.run(build_cmd, check=True)

                module_id = tool_name.replace("tool_", "", 1)
                serialized = next(
                    (p for p in self.serialized if os.path.splitext(os.path.basename(p))[0] == module_id),
                    None,
                )
                subprocess.run([loader, tool_lib, serialized or ""], check=True)


if __name__ == "__main__":
    unittest.main()

File: tests/tool_dll_test.cpp
#include <filesystem>
#include <fstream>
#include <iostream>
#include <random>
#include <set>
#include <string>
#include <vector>

#include "module_library_actualizer.h"
#include "plugin_loader.h"
#include "tool_api.h"
#include "tool_registry.h"

namespace fs = std::filesystem;

class StubHost final : public HostAPI {
public:
    void log(const char* message) override {
        if (message) std::cerr << message << "\n";
    }
    void draw_text(const TextRenderArgs& /*args*/) override {}
};

static std::string make_temp_name(const char* prefix) {
    std::random_device rd;
    std::mt19937_64 gen(rd());
    uint64_t v = gen();
    return std::string(prefix) + std::to_string(v);
}

static std::vector<fs::path> list_files_with_ext(const fs::path& root, const char* ext) {
    std::vector<fs::path> out;
    if (!fs::exists(root)) return out;
    for (auto const& entry : fs::directory_iterator(root)) {
        if (!entry.is_regular_file()) continue;
        if (entry.path().extension() == ext) out.push_back(entry.path());
    }
    return out;
}

static fs::path find_canvas_import_lib(const fs::path& build_dir) {
    const fs::path release = build_dir / "Release" / "canvas_tables.lib";
    if (fs::exists(release)) return release;
    const fs::path root = build_dir / "canvas_tables.lib";
    if (fs::exists(root)) return root;
    for (auto const& entry : fs::recursive_directory_iterator(build_dir)) {
        if (!entry.is_regular_file()) continue;
        if (entry.path().filename() == "canvas_tables.lib") return entry.path();
    }
    return {};
}

static fs::path find_tool_dll(const fs::path& build_dir, const std::string& tool_name) {
    const fs::path out = build_dir / "out" / (tool_name + ".dll");
    if (fs::exists(out)) return out;
    const fs::path release = build_dir / "Release" / (tool_name + ".dll");
    if (fs::exists(release)) return release;
    const fs::path root = build_dir / (tool_name + ".dll");
    if (fs::exists(root)) return root;
    for (auto const& entry : fs::recursive_directory_iterator(build_dir)) {
        if (!entry.is_regular_file()) continue;
        if (entry.path().filename() == (tool_name + ".dll")) return entry.path();
    }
    return {};
}

static bool run_command(const std::string& cmd) {
    int rc = std::system(cmd.c_str());
    return rc == 0;
}

static bool write_text_file(const fs::path& path, const std::string& contents) {
    std::ofstream ofs(path, std::ios::binary | std::ios::trunc);
    if (!ofs.good()) return false;
    ofs << contents;
    return ofs.good();
}

int main() {
#ifndef _WIN32
    std::cerr << "tool_dll_test skipped: plugin loader only implemented on Windows.\n";
    return 0;
#endif

    const fs::path repo_root = fs::path(NODUS_REPO_ROOT);
    const fs::path build_dir = fs::path(NODUS_BUILD_DIR);
    const fs::path repo_module_root = repo_root / "module_library";
    const fs::path serialized_dir = repo_module_root / "serialized";

    auto serialized = list_files_with_ext(serialized_dir, ".gpmod");
    if (serialized.empty()) {
        std::cerr << "tool_dll_test skipped: no serialized modules.\n";
        return 0;
    }

    fs::path test_root = repo_module_root / "test_build" / "tool_dll_test";
    const fs::path tool_src_dir = test_root / "source" / "tools";
    fs::create_directories(test_root);

    fs::path modlib_path = test_root / "module_library.txt";
    std::ostringstream lib;
    lib << "MODULELIB V1\n";
    lib << "ROOT \"" << test_root.generic_string() << "\"\n";
    for (size_t i = 0; i < serialized.size(); ++i) {
        const auto& ser_path = serialized[i];
        std::string module_id = ser_path.stem().string();
        std::string tool_id = "tool_" + module_id;
        lib << "MODULE " << module_id << " " << i << " \"" << module_id << "\" \""
            << ser_path.generic_string() << "\" \"\" 1 \"" << tool_id << "\" 0\n";
    }
    if (!write_text_file(modlib_path, lib.str())) {
        std::cerr << "failed to write module library file\n";
        return 1;
    }

    std::set<fs::path> preexisting;
    for (const auto& p : list_files_with_ext(tool_src_dir, ".cpp")) preexisting.insert(p);

    if (!gp_module_library_actualize_from_file(modlib_path.string().c_str(),
                                               test_root.string().c_str())) {
        std::cerr << "module library actualization failed\n";
        return 1;
    }

    std::vector<fs::path> generated;
    for (const auto& p : list_files_with_ext(tool_src_dir, ".cpp")) {
        if (preexisting.find(p) == preexisting.end()) generated.push_back(p);
    }
    if (generated.empty()) {
        std::cerr << "no tool sources generated\n";
        return 1;
    }

    fs::path canvas_lib = find_canvas_import_lib(build_dir);
    if (canvas_lib.empty()) {
        std::cerr << "canvas_tables.lib not found in build dir\n";
        return 1;
    }

    StubHost host;
    PluginLoader loader;
    bool ok = true;

    for (const auto& tool_source : generated) {
        std::string tool_name = tool_source.stem().string();
        fs::path proj_dir = test_root / ("cmake_" + tool_name);
        fs::path proj_build = proj_dir / "build";
        fs::create_directories(proj_dir);

        std::ostringstream cmake;
        cmake << "cmake_minimum_required(VERSION 3.15)\n";
        cmake << "project(" << tool_name << " LANGUAGES CXX)\n";
        cmake << "add_library(" << tool_name << " SHARED \"" << tool_source.generic_string() << "\")\n";
        cmake << "target_include_directories(" << tool_name << " PRIVATE \""
              << repo_root.generic_string() << "\" \""
              << (repo_root / "include").generic_string() << "\")\n";
        cmake << "target_link_libraries(" << tool_name << " PRIVATE \""
              << canvas_lib.generic_string() << "\")\n";
        cmake << "set_target_properties(" << tool_name << " PROPERTIES CXX_STANDARD 17)\n";
        cmake << "set(CMAKE_RUNTIME_OUTPUT_DIRECTORY \"${CMAKE_BINARY_DIR}/out\")\n";
        cmake << "set(CMAKE_LIBRARY_OUTPUT_DIRECTORY \"${CMAKE_BINARY_DIR}/out\")\n";
        cmake << "set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY \"${CMAKE_BINARY_DIR}/out\")\n";
        cmake << "set(CMAKE_PDB_OUTPUT_DIRECTORY \"${CMAKE_BINARY_DIR}/out\")\n";
        cmake << "set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_RELEASE \"${CMAKE_BINARY_DIR}/out\")\n";
        cmake << "set(CMAKE_LIBRARY_OUTPUT_DIRECTORY_RELEASE \"${CMAKE_BINARY_DIR}/out\")\n";
        cmake << "set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY_RELEASE \"${CMAKE_BINARY_DIR}/out\")\n";
        cmake << "set(CMAKE_PDB_OUTPUT_DIRECTORY_RELEASE \"${CMAKE_BINARY_DIR}/out\")\n";

        if (!write_text_file(proj_dir / "CMakeLists.txt", cmake.str())) {
            std::cerr << "failed to write CMakeLists for " << tool_name << "\n";
            ok = false;
            break;
        }

        std::string cfg_cmd = "cmake -S \"" + proj_dir.string() + "\" -B \"" +
                              proj_build.string() + "\" -DCMAKE_BUILD_TYPE=Release";
        std::string build_cmd = "cmake --build \"" + proj_build.string() + "\" --config Release";
        if (!run_command(cfg_cmd) || !run_command(build_cmd)) {
            std::cerr << "failed to build tool " << tool_name << "\n";
            ok = false;
            break;
        }

        fs::path tool_lib = find_tool_dll(proj_build, tool_name);
        if (tool_lib.empty()) {
            std::cerr << "tool DLL not found in: " << proj_build.string() << "\n";
            ok = false;
            break;
        }

        std::string id = loader.load_module(tool_lib.string(), &host);
        if (id.empty()) {
            std::cerr << "failed to load tool DLL: " << tool_lib.string() << "\n";
            ok = false;
            break;
        }

        auto tool = tool_registry_global().create(id);
        if (!tool) {
            std::cerr << "failed to create tool instance for: " << id << "\n";
            ok = false;
            break;
        }
        ToolInitContext ctx{};
        RenderContext rctx{};
        tool->initialize(ctx);
        tool->tick(0.016, host);
        tool->render(rctx);
        tool->shutdown();
        tool.reset();
        loader.unload_module(id);
    }

    return ok ? 0 : 1;
}
